astropy/units/core.py

```diff
@@ -2094,14 +2094,24 @@ def __call__(
             if isinstance(s, bytes):
                 s = s.decode("ascii")
 
+            warning_actions = {"silent": "ignore", "warn": "default", "raise": "error"}
             try:
-                return f.parse(s)
+                warning_action = warning_actions[parse_strict]
+            except KeyError:
+                raise ValueError(
+                    "'parse_strict' must be 'warn', 'raise' or 'silent'"
+                ) from None
+            try:
+                with warnings.catch_warnings(
```

Comment:
Looking at the [docs](https://docs.python.org/3/library/warnings.html#warnings.catch_warnings) I see that `catch_warnings` isn't thread-safe. Was the previous implementation?

Reply:
We've had quite a few PRs dealing with thread-safety adding locks, etc. So, my guess is that, as is, we were thread-safe (or at least "safe enough"). It also seems this is a real problem: the risk is that a thread gets our changed state, i.e., if we use `parse_strict="raise"` (the default), it would suddenly start failing on all warnings after unit parsing...

From https://stackoverflow.com/questions/56043644/python-locally-suppress-warnings-in-thread-safe-way, it looks like the only solution is to add a lock. That's not very complicated and adds negligible overhead compared to the cost we have anyway in setting up the warning filter (0.2 us vs 1.9 us). So, that seems the simplest solution.

However, seeing the time to set up `catch_warnings` makes me wonder again if we cannot do better another way, since it is a significant overhead, more than doubling the parsing time for simple units (on the same laptop, `u.Unit("m")` takes 1.5 us currently).

I guess that gets us back to whether we should just bite the bullet and pass on `parse_strict` after all. Though in a way that just pushes the problem to the parser: the warnings are raised in the middle of the action, inside bits that only have access to bits of the string, so one still needs a process of letting any warnings bubble up anyway, i.e., something like a lock (but at least it would not increase the overhead of the generic parser!). The same would be the case if we implement something that, say, attaches an attribute to the returning unit of whether there was a warning during parsing.

The other way this perhaps can be approached is to have a utility function or method for the warnings, i.e., we call the parser with, effectively, our own `catch_warnings` (with a lock), and the parser has to use the units utility function (say, via a new `cls.warn(...)` method), which would get the locked units warning item, and add something to it. I think this may be possible to implement this relatively easily with a [context variable](https://docs.python.org/3/library/contextvars.html). (Don't really know details, but became aware of it via https://github.com/numpy/numpy/pull/26846).

What do you both think?

Reply:
I've opened #17004 to speed up parsing simple unit strings with `Unit`. If that is merged then this pull request wouldn't slow down parsing simple strings at all. It would slightly slow down parsing composite units, but those are slower to much parse anyways, so the relative slowdown would be small.

I still have to think about thread safety.

Reply:
Sounds good! It keeps things simple from the parser perspective, which was always the big incentive for the approach here.

Reply:
context variables are cool! I didn't know about them. Thanks 👍

Reply:
Just saw this because of the backlink from the NumPy PR. I think these days you can think of context variables as a "modern" async-aware (but still thread-safe) version of `threading.local`. I also just learned today that `gevent` [treats context variables as locals](https://www.gevent.org/api/gevent.contextvars.html) to a green thread, so even async frameworks outside of the built-in ones use them now. You get lots of parallel/concurrent safety "for free" if you put state that could be stored in a global in a context variable.